src/pb_stub.cc

```diff
@@ -1115,13 +1122,36 @@ Stub::ParentToStubServiceActive()
   return parent_to_stub_thread_;
 }
 
-void
-Stub::SaveResponseIterator(std::shared_ptr<ResponseIterator> response_iterator)
+std::shared_ptr<ResponseIterator>
+Stub::GetResponseIterator(std::shared_ptr<InferResponse> infer_response)
 {
   std::lock_guard<std::mutex> lock(response_iterator_map_mu_);
-  response_iterator_map_.insert(
-      std::pair<void*, std::shared_ptr<ResponseIterator>>(
-          response_iterator->Id(), response_iterator));
+  if (response_iterator_map_.find(infer_response->Id()) !=
+      response_iterator_map_.end()) {
+    // Need to re-construct the 'ResponseIterator' and update the
+    // 'response_iterator_map_' to make sure the 'ResponseIterator' object has
+    // the correct first response.
+    auto response_iterator = std::make_shared<ResponseIterator>(infer_response);
+    bool done = false;
+    while (!done) {
```

Comment:
Can we make `done` an argument of `Pop` instead of raising an exception? I think we shouldn't raise Python exceptions for C++ code.

Reply:
Would changing to `PythonBackendException` help? I agree that we shouldn't be throwing `py::` object.

Reply:
I think it might be best to add a new boolean to `Pop(done)`. If `done` is set to true, we will no longer call Pop. I think exceptions should be reserved for errors etc.